fix: @SchemaSwap can only be used once

xRodney opened the pull request:
## Description

@SchemaSwap is now repeatable, so multiple modifications can be applied from the same resource root.
Fixed a bug when the swap was "used up" after just one usage, meaning that if the referenced type was used multiple times in object hierarchy, only the first one was swapped.

Fixes fabric8io/kubernetes-client#4350

<!--
Thank a lot for taking time to contribute to Fabric8 <3!

Please provide a description of what your PR does providing a link (if applicable) to the issue it fixes. It is
really helpful for people who would review your code.
-->

## Type of change
<!---
What types of changes does your code introduce? Put an `x` in all the boxes that apply
-->
 - [x] Bug fix (non-breaking change which fixes an issue)
 - [ ] Feature (non-breaking change which adds functionality)
 - [ ] Breaking change (fix or feature that would cause existing functionality to change
 - [ ] Chore (non-breaking change which doesn't affect codebase;
   test, version modification, documentation, etc.)

## Checklist
 - [x] Code contributed by me aligns with current project license: [Apache 2.0](https://www.apache.org/licenses/LICENSE-2.0)
 - [x] I Added [CHANGELOG](https://github.com/fabric8io/kubernetes-client/blob/master/CHANGELOG.md) entry regarding this change
 - [x] I have implemented unit tests to cover my changes
 - [ ] I have added/updated the [javadocs](https://www.javadoc.io/doc/io.fabric8/kubernetes-client/latest/index.html) and other [documentation](https://github.com/fabric8io/kubernetes-client/blob/master/doc/CHEATSHEET.md) accordingly
 - [ ] No new bugs, code smells, etc. in [SonarCloud](https://sonarcloud.io/dashboard?id=fabric8io_kubernetes-client) report
 - [ ] I tested my code in Kubernetes
 - [ ] I tested my code in OpenShift

<!--
Integration tests (https://github.com/fabric8io/kubernetes-client/tree/master/kubernetes-itests)
Please check integration tests and provide/improve tests if applicable.

Open your PR in Draft mode and verify all of the applicable Checklist items before marking your pull request as ready for review
-->

xRodney commented on crd-generator/api/src/test/java/io/fabric8/crd/generator/v1/JsonSchemaTest.java:
```diff
@@ -178,15 +179,58 @@ void shouldExtractPropertiesSchemaFromExtractValueAnnotation() {
+    assertEquals(
+        "Unmatched SchemaSwaps: @SchemaSwap(originalType=io.fabric8.crd.example.extraction.ExtractionSpec, fieldName=\"FOO\", targetType=io"
+            + ".fabric8.crd.example.extraction.FooExtractor) on io.fabric8.crd.example.extraction.IncorrectExtraction",
+        exception.getMessage());
```
The new implementation no longer differentiates between unmatched class and unmatched field. I my opinion the more detailed message compensates for it.

andreaTP replied:
totally fine to change the message, but I believe that this change of behavior is caused by the fact that we are missing an invariant, see the comment above on matching the field.

andreaTP commented on crd-generator/api/src/main/java/io/fabric8/crd/generator/AbstractJsonSchema.java:
```diff
@@ -239,11 +256,9 @@ private T internalFromImpl(TypeDef definition, Set<String> visited, List<Interna
       }
 
-      final PropertyFacade facade = new PropertyFacade(property, accessors, currentSchemaSwaps);
+      ClassRef potentialSchemaSwap = schemaSwaps.lookupAndMark(definition.toReference(), name).orElse(null);
```
If I correctly follow this change, here you cannot mark just yet the SchemaSwap as used, as it will mark the annotation to be "used" even in this condition:
 - Matched `Class`
 - The Class doesn't have the specified `Property` (e.g. a matching field)
 
 This operation should be performed only after `process` has been called on the `PropertyFacade` (and the `SchemaSwap` has been matched).

xRodney replied:
schemaSwapt.lookupAndMark  accepts two arguments, the second being the property name. (Line 253: `String name = property.getName();`)
So I'm not sure I understand your concert?
Btw, I thought the case of "matched class, unmatched field" is being tested in JsonSchemaTest.shouldThrowIfSchemaSwapHasUnmatchedField. This test is still passing and I only extended it with assert on the error message. But the error is still thrown.

xRodney replied:
Of is your concern that the `name` could be overriden by `@JsonProperty` & friends? IMHO it makes more sense for the `fieldName` parameter to reference the fields as is defined in Java, but that's just my opinion.

andreaTP replied:
> it makes more sense for the fieldName parameter to reference the fields as is defined in Java, but that's just my opinion.

This makes sense, but, using `name` might not be the name extracted by the `PropertyFacade` constructor:
https://github.com/fabric8io/kubernetes-client/blob/808084e0e348d5b377ac31fd89a979996867d93e/crd-generator/api/src/main/java/io/fabric8/crd/generator/AbstractJsonSchema.java#L414-L426

in some cases, the `PropertyFacade` is extracting the field name from constructor/getter/setter 
and this implementation will not respect the extracted name as opposed to the underlying field name.
We can either refactor the logic and use the `PropertyFacade` name extraction logic everywhere or simply document this different behavior.

xRodney replied:
I see, and the only way how that can happen is a `@JsonProperty` on the field itself or getter (get* or is*) or setter. 
I can only say that to me, as the user of the library, would never occur that `fieldName` could reference the name after transformations. 
Maybe it's because the word "field" is used, whose meaning in Java is well defined and is different from the term "property". (Field is a class variable, property is a getter/setter pair - which often is backed by a field). Also the annotation is `@JsonProperty`, not "JsonField".
Or maybe it's because the reason I am dealing with `@SchemaSwaps` in the first place is that I cannot place annotations (like `@JsonProperty`) on the field directly.
So with your permission, I will add Javadocs on `@SchemaSwap`, describing the currently implemented behaviour.

andreaTP replied:
To me, this "limitation" is totally fine, please just make it explicit in this doc:
https://github.com/fabric8io/kubernetes-client/blob/master/doc/CRD-generator.md#iofabric8crdgeneratorannotationschemaswap

mentioning something along the lines:
> the name of the field is restricted to the original `fieldName` and should be backed by a matching concrete field of the matching class. Getters, setters, and constructors are not taken into consideration.

cc. @metacosm to check if he is on board with this decision.

metacosm replied:
I think that should be fine but this should be documented as a breaking change, though…

andreaTP replied:
I agree that it "might" be breaking, but, at the same time, this PR is solving some fundamental issues with the implementation and I doubt anyone else is using it just yet.

andreaTP replied:
Maybe we can add a line to the Changelog for reference, wdyt?

andreaTP commented on crd-generator/api/src/main/java/io/fabric8/crd/generator/InternalSchemaSwaps.java:
```diff
@@ -0,0 +1,133 @@
+    }
+  }
+
+  public static class Value {
```
This `Value` class spills out, either we give it a more meaningful name, or we contain it making it `private` and returning a more generic type from `getUnusedSwaps` (e.g. directly the resulting `String` to be used in the error message?)

Also,  given this implementation, `Value` can extend `Key`, I believe.

xRodney replied:
I renamed it to `InternalSchemaSwap`.
About extending `Key`, yeah, it's a possibility, but I would rather not do it, because `Key` is an immutable value class, designed to be used as, well, a HashMap key. That has implications on it's `equals` and `hashCode`.
`Value` (or now `InternalSchemaSwap`) is mutable, so either its `equals/hashCode` would have to only use a subset of fields or it could never be used as a map key - but being a subclass of `Key`, any `HashMap<Key, ...>` would allow that. So I chose to not deal with this at all and make it separate.

andreaTP replied:
fair, but mark `Key` as `final` please.

andreaTP commented on crd-generator/api/src/test/java/io/fabric8/crd/generator/v1/JsonSchemaTest.java:
```diff
@@ -178,15 +179,58 @@ void shouldExtractPropertiesSchemaFromExtractValueAnnotation() {
   }
 
+  @Test
+  void shouldExtractPropertiesSchemaFromSchemaSwapAnnotations() {
```
what's the difference in between this test and the one extracted in `SchemaSwapSpec` ? Or is it a leftover?

xRodney replied:
There are two very similar tests: this one in crd-generator/api project, and another in crd-generator/test. 
They both perform the same set of modifications on the same schema, but the difference is that one runs in reflection context and the other on schema generated by Apt. 
In an ideal world, that shouldn't matter, but the sundr library is not yet perfect with annotation arguments that themselves are annotations, and these are sometimes presented as AnnotationRef and sometimes directly as the target annotation (SchemaSwap). Therefore, the code in AbstractJsonSchema needs to compensate for this by using two code paths, which both need to be tested.

andreaTP reviewed (changes requested):
@xRodney Thanks a lot for this PR!
I really like the overall changes that you did 🙂 , I added a few notes to be discussed.

A few additional requests, since you are fixing a bug can you add test cases for the relevant situations where the current implementation is failing (what you already described in the issue):
- `SchemaSwap` should be applied to all occurrences in the hierarchy
- `SchemaSwap` should be applied to deeply nested structures (e.g. 3 or more levels down)
- `SchemaSwap` should fail if the field name is not matched

xRodney commented:
I have pushed one more test for deeply nested hierarchies, which also tests that the Swap is applied multiple times. 
As for the third case you mentioned, when name is not matched, I believe that case is already covered by a pre-existing test, but let's discuss that in that other thread.

andreaTP reviewed (approved):
Thanks again for this effort @xRodney !
LGTM, but let's wait for a little to see if @metacosm wanna chip in.

andreaTP commented:
@xRodney can you please fix the issues with the CI? `mvn spotless:apply -f crd-generator/pom.xml` should make it automatically.

xRodney commented:
Hey @andreaTP, thanks to you, too for your help.
I did the reformat and also at the last minute I noticed I forgot to delete the now unused `InternalSchemaSwap` inner class. And that led to realization that I can move the logic of `validateRemainingSchemaSwaps` into the new class `InternalSchemaSwaps`, so its inner class no longer leaks out and can be renamed back to `Value`.
And last but not least, I squashed everything into one commit.
So I'll wait for the CI to finish and for @metacosm's review and we're ready to take off.

xRodney commented:
Hi @andreaTP, it appears that you need to run the CI for me, with me being a first-time contributor? If so, could you please do it, so that we know if we are in the green?

xRodney commented:
I have updated the changelog and added the breaking change as requested. In the latest iteration I have also added a few more tests about the case when targetType is void. And also rebased.
Please, run the CI one more time and if it passes, I think we are ready for merge.

andreaTP commented:
@xRodney we merged another PR slightly conflicting with the changes in this PR.
Would you mind rebasing? Please keep your JavaDoc style as it's much better 🙂 

Ping me and I will run the CI to validate and eventually move this forward 👍 thanks a lot for your effort!

xRodney commented:
Sure, it's done

andreaTP commented:
Unfortunately, the CI failure seems to be legit, @xRodney can you have a look?

xRodney commented:
Oops, missing import. Sorry about that, it's fixed now.

andreaTP reviewed (approved):
LGTM, thanks!

andreaTP commented:
thanks @xRodney !
